File: AlbumLibrary/src/main/java/com/album/ui/adapter/AlbumAdapter.java
package com.album.ui.adapter;

import android.graphics.PorterDuff;
import android.graphics.drawable.Drawable;
import android.support.annotation.NonNull;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.AppCompatCheckBox;
import android.support.v7.widget.AppCompatImageView;
import android.support.v7.widget.AppCompatTextView;
import android.support.v7.widget.RecyclerView;
import android.text.TextUtils;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;

import com.album.Album;
import com.album.AlbumConfig;
import com.album.AlbumConstant;
import com.album.R;
import com.album.entity.AlbumEntity;
import com.album.ui.widget.AlbumImageView;
import com.album.util.FileUtils;

import java.util.ArrayList;
import java.util.Objects;

/**
 * by y on 14/08/2017.
 */

public class AlbumAdapter extends RecyclerView.Adapter<AlbumAdapter.ViewHolder> {


    private ArrayList<AlbumEntity> albumList;
    private ArrayList<AlbumEntity> multiplePreviewList = null;
    private OnItemClickListener onItemClickListener = null;
    private final AlbumConfig albumConfig;
    private final int display;
    private final FrameLayout.LayoutParams layoutParams;

    public AlbumAdapter(ArrayList<AlbumEntity> list, int display) {
        this.albumList = list;
        this.display = display;
        albumConfig = Album.getInstance().getConfig();
        layoutParams = new FrameLayout.LayoutParams(display, display);
        if (!albumConfig.isRadio()) {
            multiplePreviewList = new ArrayList<>();
        }
    }


    public void setOnItemClickListener(OnItemClickListener onItemClickListener) {
        this.onItemClickListener = onItemClickListener;
    }

    @NonNull
    @Override
    public AlbumAdapter.ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        View view;
        view = LayoutInflater.from(parent.getContext()).inflate(R.layout.item_album, parent, false);
        final ViewHolder viewHolder = new ViewHolder(view);
        view.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (onItemClickListener != null && albumList != null) {
                    onItemClickListener.onItemClick(v, viewHolder.getAdapterPosition(), albumList.get(viewHolder.getAdapterPosition()));
                }
            }
        });
        return viewHolder;
    }

    @Override
    public void onBindViewHolder(@NonNull final AlbumAdapter.ViewHolder holder, int position) {

        if (albumList == null) {
            return;
        }
        final AlbumEntity albumEntity = albumList.get(position);
        if (albumEntity == null) {
            return;
        }
        String path = albumEntity.getPath();
        if (TextUtils.equals(String.valueOf(path), AlbumConstant.CAMERA)) {
            holder.camera();
        } else {
            holder.cameraRootView.setVisibility(View.GONE);
            holder.imageView.setVisibility(View.VISIBLE);
            ImageView imageView;
            if (albumConfig.isFrescoImageLoader()) {
                imageView = Album.getInstance().getAlbumImageLoader().frescoView(holder.imageView.getContext(), AlbumConstant.TYPE_FRESCO_ALBUM);
            } else {
                imageView = new AlbumImageView(holder.imageView.getContext());
            }
            imageView.setLayoutParams(layoutParams);
            holder.imageView.addView(imageView);
            Album.getInstance().getAlbumImageLoader().displayAlbum(imageView, display, display, albumEntity);
            if (!albumConfig.isRadio()) {
                holder.checkBox.setVisibility(View.VISIBLE);
                holder.checkBox.setChecked(albumEntity.isCheck());
                holder.checkBox.setBackgroundResource(albumConfig.getAlbumContentItemCheckBoxDrawable());
                holder.checkBox.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View v) {
                        if (!FileUtils.isFile(albumEntity.getPath())) {
                            holder.checkBox.setChecked(false);
                            Album.getInstance().getAlbumListener().onAlbumCheckBoxFileNull();
                            return;
                        }
                        if (!multiplePreviewList.contains(albumEntity) && multiplePreviewList.size() >= albumConfig.getMultipleMaxCount()) {
                            holder.checkBox.setChecked(false);
                            Album.getInstance().getAlbumListener().onAlbumMaxCount();
                            return;
                        }
                        if (!albumEntity.isCheck()) {
                            albumEntity.setCheck(true);
                            multiplePreviewList.add(albumEntity);
                        } else {
                            multiplePreviewList.remove(albumEntity);
                            albumEntity.setCheck(false);
                        }
                    }
                });
            }
        }
    }

    @Override
    public int getItemCount() {
        return albumList == null ? 0 : albumList.size();
    }

    public void addAll(ArrayList<AlbumEntity> list) {
        if (albumList == null) {
            albumList = new ArrayList<>();
        }
        if (list != null && !list.isEmpty()) {
            albumList.addAll(list);
            notifyDataSetChanged();
        }
    }

    public ArrayList<AlbumEntity> getAlbumList() {
        return albumList;
    }

    public ArrayList<AlbumEntity> getMultiplePreviewList() {
        return multiplePreviewList;
    }

    public void setMultiplePreviewList(ArrayList<AlbumEntity> multiplePreviewList) {
        this.multiplePreviewList = multiplePreviewList;
        notifyDataSetChanged();
    }

    public void removeAll() {
        if (albumList != null) {
            albumList.clear();
        }
        notifyDataSetChanged();
    }

    public interface OnItemClickListener {
        void onItemClick(View view, int position, AlbumEntity albumEntity);
    }

    class ViewHolder extends RecyclerView.ViewHolder {
        private final FrameLayout imageView;
        private final AppCompatCheckBox checkBox;
        private final AppCompatImageView imageCamera;
        private final AppCompatTextView cameraTips;
        private final LinearLayout cameraRootView;

        ViewHolder(View itemView) {
            super(itemView);
            imageView = itemView.findViewById(R.id.album_image);
            checkBox = itemView.findViewById(R.id.album_check_box);
            imageCamera = itemView.findViewById(R.id.album_image_camera);
            cameraTips = itemView.findViewById(R.id.album_image_camera_tv);
            cameraRootView = itemView.findViewById(R.id.album_camera_root_view);
        }

        void camera() {
            Drawable drawable = ContextCompat.getDrawable(itemView.getContext(), albumConfig.getAlbumContentViewCameraDrawable());
            Objects.requireNonNull(drawable).setColorFilter(ContextCompat.getColor(itemView.getContext(), albumConfig.getAlbumContentViewCameraDrawableColor()), PorterDuff.Mode.SRC_ATOP);
            cameraTips.setText(albumConfig.getAlbumContentViewCameraTips());
            cameraTips.setTextSize(albumConfig.getAlbumContentViewCameraTipsSize());
            cameraTips.setTextColor(ContextCompat.getColor(itemView.getContext(), albumConfig.getAlbumContentViewCameraTipsColor()));
            cameraRootView.setBackgroundColor(ContextCompat.getColor(itemView.getContext(), albumConfig.getAlbumContentViewCameraBackgroundColor()));
            imageCamera.setImageDrawable(drawable);
            cameraRootView.setVisibility(View.VISIBLE);
            imageView.setVisibility(View.GONE);
            checkBox.setVisibility(View.GONE);
        }
    }
}

File: AlbumLibrary/src/main/java/com/album/ui/adapter/AlbumAdapter.kt
package com.album.ui.adapter

import android.graphics.PorterDuff
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.LinearLayout
import androidx.appcompat.widget.AppCompatCheckBox
import androidx.appcompat.widget.AppCompatImageView
import androidx.appcompat.widget.AppCompatTextView
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.RecyclerView
import com.album.Album
import com.album.AlbumBundle
import com.album.AlbumInternalConst
import com.album.R
import com.album.core.fileExists
import com.album.core.scan.AlbumEntity
import com.album.core.show
import com.album.listener.AlbumCallback
import com.album.listener.OnAlbumItemClickListener
import com.album.listener.addChildView

class AlbumAdapter(
        private val display: Int,
        private val albumBundle: AlbumBundle,
        private val albumCallback: AlbumCallback?,
        private val onAlbumItemClickListener: OnAlbumItemClickListener
) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {

    companion object {
        private const val TYPE_CAMERA = 0
        private const val TYPE_PHOTO = 1
    }

    var albumList: ArrayList<AlbumEntity> = ArrayList()

    var multipleList: ArrayList<AlbumEntity> = ArrayList()
        set(value) {
            field = value
            notifyDataSetChanged()
        }

    private val layoutParams: FrameLayout.LayoutParams = FrameLayout.LayoutParams(display, display)

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): RecyclerView.ViewHolder {
        return when (viewType) {
            TYPE_CAMERA -> {
                val cameraView: View = LayoutInflater.from(parent.context).inflate(R.layout.album_item_album_camera, parent, false)
                val cameraViewHolder = CameraViewHolder(cameraView, albumBundle)
                cameraView.setOnClickListener { v -> onAlbumItemClickListener.onCameraItemClick(v, cameraViewHolder.adapterPosition, albumList[cameraViewHolder.adapterPosition]) }
                cameraViewHolder
            }
            else -> {
                val photoView: View = LayoutInflater.from(parent.context).inflate(R.layout.album_item_album, parent, false)
                val photoViewHolder = PhotoViewHolder(photoView, albumBundle, display, layoutParams, albumCallback)
                photoView.setOnClickListener { v -> onAlbumItemClickListener.onPhotoItemClick(v, photoViewHolder.adapterPosition, albumList[photoViewHolder.adapterPosition]) }
                photoViewHolder
            }
        }
    }

    override fun onBindViewHolder(holder: RecyclerView.ViewHolder, position: Int) {
        if (holder is CameraViewHolder) {
            holder.camera()
        } else if (holder is PhotoViewHolder) {
            holder.photo(position, albumList[position], multipleList)
        }
    }

    override fun getItemCount(): Int = albumList.size

    override fun getItemViewType(position: Int): Int = when {
        albumList.isEmpty() -> TYPE_PHOTO
        albumList[position].path == AlbumInternalConst.CAMERA -> TYPE_CAMERA
        else -> TYPE_PHOTO
    }

    fun addAll(newList: ArrayList<AlbumEntity>) {
        albumList.addAll(newList)
        notifyDataSetChanged()
    }

    fun removeAll() {
        albumList.clear()
        notifyDataSetChanged()
    }

    class CameraViewHolder(itemView: View, private val albumBundle: AlbumBundle) : RecyclerView.ViewHolder(itemView) {

        private val container: LinearLayout = itemView.findViewById(R.id.album_camera_root_view)
        private val cameraIv: AppCompatImageView = itemView.findViewById(R.id.albumImageCamera)
        private val cameraTv: AppCompatTextView = itemView.findViewById(R.id.albumImageCameraTv)

        fun camera() {
            val drawable = ContextCompat.getDrawable(itemView.context, albumBundle.cameraDrawable)
            drawable?.setColorFilter(ContextCompat.getColor(itemView.context, albumBundle.cameraDrawableColor), PorterDuff.Mode.SRC_ATOP)
            cameraTv.setText(albumBundle.cameraText)
            cameraTv.textSize = albumBundle.cameraTextSize
            cameraTv.setTextColor(ContextCompat.getColor(itemView.context, albumBundle.cameraTextColor))
            container.setBackgroundColor(ContextCompat.getColor(itemView.context, albumBundle.cameraBackgroundColor))
            cameraIv.setImageDrawable(drawable)
        }
    }

    class PhotoViewHolder(itemView: View,
                          private val albumBundle: AlbumBundle,
                          private val display: Int,
                          private val layoutParams: ViewGroup.LayoutParams,
                          private val callback: AlbumCallback?) : RecyclerView.ViewHolder(itemView) {

        private val container: FrameLayout = itemView.findViewById(R.id.albumContainer)
        private val checkBox: AppCompatCheckBox = itemView.findViewById(R.id.albumCheckBox)

        fun photo(position: Int, albumEntity: AlbumEntity, multipleList: ArrayList<AlbumEntity>) {
            container.addChildView(Album.instance.albumImageLoader?.displayAlbum(display, display, albumEntity, container), layoutParams)
            container.setBackgroundColor(ContextCompat.getColor(itemView.context, albumBundle.photoBackgroundColor))
            if (albumBundle.radio) {
                return
            }
            checkBox.show()
            checkBox.isChecked = albumEntity.isCheck
            checkBox.setBackgroundResource(albumBundle.checkBoxDrawable)
            checkBox.setOnClickListener {
                if (callback?.onAlbumCheckBoxFilter(itemView, position, albumEntity) == true) {
                    return@setOnClickListener
                }
                if (!albumEntity.path.fileExists()) {
                    checkBox.isChecked = false
                    Album.instance.albumListener?.onAlbumCheckFileNotExist()
                    return@setOnClickListener
                }
                if (!multipleList.contains(albumEntity) && multipleList.size >= albumBundle.multipleMaxCount) {
                    checkBox.isChecked = false
                    Album.instance.albumListener?.onAlbumMaxCount()
                    return@setOnClickListener
                }
                if (!albumEntity.isCheck) {
                    albumEntity.isCheck = true
                    multipleList.add(albumEntity)
                } else {
                    multipleList.remove(albumEntity)
                    albumEntity.isCheck = false
                }
                callback?.onChangedCheckBoxCount(itemView, multipleList.size, albumEntity)
                Album.instance.albumListener?.onAlbumCheckBox(multipleList.size, albumBundle.multipleMaxCount)
            }
        }
    }
}
